File: leetcode/algorithms/5_longest_palindromic_substring/main.js
/**
 * Expand Around Center
 * - Time Complexity: O(N²)
 * - Space Complexity: O(1)
 */
/**
 * @param {string} s
 * @param {number} left
 * @param {number} right
 * @return {number}
 */
var getPalindromeLength = function (s, left, right) {
  while (left >= 0 && right < s.length && s[left] === s[right]) {
    left--;
    right++;
  }
  return right - left - 1;
};

/**
 * @param {string} s
 * @return {string}
 */
var longestPalindrome = function (s) {
  if (s.length === 1) {
    return s;
  }

  let startIndex = 0;
  let maxLen = 0;

  for (let i = 0; i < s.length; i++) {
    const oddLen = getPalindromeLength(s, i, i);
    const evenLen = getPalindromeLength(s, i, i + 1);

    const currMaxLen = oddLen > evenLen ? oddLen : evenLen;

    if (currMaxLen > maxLen) {
      maxLen = currMaxLen;
      startIndex = i - Math.floor((maxLen - 1) / 2);
    }
  }

  return s.substring(startIndex, startIndex + maxLen);
};


// Solution
/**
 * Solution 1
 *
 * Expand Around Center
 * - Time Complexity: O(N²)
 * - Space Complexity: O(1)
 */
/**
 * @param {string} s
 * @return {string}
 */
var solution1 = function (s) {
  if (s.length < 2) return s;

  let start = 0;
  let maxLength = 1;

  /**
   * @param {number} left - 시작 왼쪽 포인터
   * @param {number} right - 시작 오른쪽 포인터
   */
  function expandAroundCenter(left, right) {
    // 포인터가 유효 범위 내에 있고, 양쪽 문자가 같을 동안 확장
    while (left >= 0 && right < s.length && s[left] === s[right]) {
      const currentLength = right - left + 1;
      if (currentLength > maxLength) {
        maxLength = currentLength;
        start = left;
      }
      left--;
      right++;
    }
  }

  for (let i = 0; i < s.length; i++) {
    // 홀수 길이 Palindrome (중심이 i)
    expandAroundCenter(i, i);
    // 짝수 길이 Palindrome (중심이 i, i+1)
    expandAroundCenter(i, i + 1);
  }

  return s.substring(start, start + maxLength);
};

/**
 * Solution 2
 *
 * Dynamic Programming
 * - Time Complexity: O(N²)
 * - Space Complexity: O(N²)
 */
/**
 * @param {string} s
 * @return {string}
 */
var solution2 = function (s) {
  if (s.length < 2) {
    return s;
  }

  const n = s.length;
  const dp = Array.from({ length: n }, () => new Array(n).fill(false));

  let start = 0;
  let maxLength = 1;

  // 길이가 1인 경우
  for (let i = 0; i < n; i++) {
    dp[i][i] = true;
  }

  // 길이가 2인 경우
  for (let i = 0; i < n - 1; i++) {
    if (s[i] === s[i + 1]) {
      dp[i][i + 1] = true;
      start = i;
      maxLength = 2;
    }
  }

  // 길이가 3 이상인 경우
  for (let len = 3; len <= n; len++) {
    for (let i = 0; i <= n - len; i++) {
      const j = i + len - 1;

      if (s[i] === s[j] && dp[i + 1][j - 1]) {
        dp[i][j] = true;

        if (len > maxLength) {
          start = i;
          maxLength = len;
        }
      }
    }
  }

  return s.substring(start, start + maxLength);
};

File: leetcode/algorithms/5_longest_palindromic_substring/main.py
from math import floor


class LongestPalindromicSubstring:
    """
    # Expand Around Center
    # 
    # Complexities:
    #   N - Length of `s`
    #   - Time Complexity: O(N²)
    #   - Space Complexity: O(1)
    """
    def getPalindromeLength(self, s: str, s_len: int, left: int, right: int):
        while left >= 0 and right < s_len and s[left] == s[right]:
            left -= 1
            right += 1
        return right - left - 1

    def longestPalindrome(self, s: str) -> str:
        s_len: int = len(s)

        if s_len == 1:
            return s

        start_index: int = 0
        max_len: int = 0

        for i in range(0, s_len):
            odd_len: int = self.getPalindromeLength(s, s_len, i, i)
            even_len: int = self.getPalindromeLength(s, s_len, i, i + 1)

            curr_max_len: int = max(odd_len, even_len)

            if curr_max_len > max_len:
                max_len = curr_max_len
                start_index = i - floor((max_len - 1) / 2)

        return s[start_index: start_index + max_len]


    # Solution
    """
    # Solution 1
    #
    # Expand Around Center
    # 
    # Complexities:
    #   N - Length of `s`
    #   - Time Complexity: O(N²)
    #   - Space Complexity: O(1)
    """
    def longestPalindrome(self, s: str) -> str:
        if len(s) < 2 or s == s[::-1]:
            return s

        start, max_len = 0, 1

        def expand(left: int, right: int) -> int:
            # left와 right가 유효한 인덱스이고, 문자가 동일한 동안 확장
            while left >= 0 and right < len(s) and s[left] == s[right]:
                left -= 1
                right += 1

            # Palindrome의 길이는 right - left - 1
            return right - left - 1

        for i in range(len(s)):
            # 홀수 길이 Palindrome (중앙이 i)
            len1 = expand(i, i)

            # 짝수 길이 Palindrome (중앙이 i와 i+1 사이)
            len2 = expand(i, i + 1)

            current_len = max(len1, len2)

            if current_len > max_len:
                max_len = current_len

                # 시작점 계산: i - (길이-1)//2
                start = i - (current_len - 1) // 2

        return s[start: start + max_len]

    """
    # Solution 2
    # 
    # Dynamic Programming
    # 
    # Complexities:
    #   N - Length of `s`
    #   - Time Complexity: O(N²)
    #   - Space Complexity: O(N²)
    """
    def solution2(self, s: str) -> str:
        n = len(s)

        if n < 2:
            return s

        dp = [[False] * n for _ in range(n)]
        start, max_len = 0, 1

        # 길이가 1인 모든 부분 문자열은 Palindrome
        for i in range(n):
            dp[i][i] = True

        # 길이가 2인 부분 문자열 검사
        for i in range(n - 1):
            if s[i] == s[i + 1]:
                dp[i][i + 1] = True
                start = i
                max_len = 2

        # 길이가 3 이상인 부분 문자열 검사
        for length in range(3, n + 1):
            for i in range(n - length + 1):
                j = i + length - 1

                if s[i] == s[j] and dp[i + 1][j - 1]:
                    dp[i][j] = True

                    if length > max_len:
                        start = i
                        max_len = length

        return s[start: start + max_len]
